scipy/special/cephes/stirling.c

```diff
@@ -0,0 +1,54 @@
+/* c implementation of Stirling numbers of the second kind */
+
+/*
+ *
+ *     Stirling numbers of the second kind
+ *
+ *
+ *
+ * SYNOPSIS: Stirling numbers of the second kind count the
+ *  number of ways to make a partition of n distinct elements
+ *  into k non-empty subsets.
+ *
+ * DESCRIPTION: n is the number of distinct elements of the set
+ *  to be partitioned and k is the number of non-empty subsets.
+ *  The values for n < 0 or k < 0 are interpreted as 0. If you
+ *
+ * ACCURACY: The method returns an unsigned integer type and
+ *  the size of the type 32 bits.
+ *
+ *
+ * NOTE: this file is *NOT* part of the ceph distribution and was
+ *  added by Lucas Roberts in 2023 after opening this issue
+ *  https://github.com/scipy/scipy/issues/17890
+ */
+#include "mconf.h"
+#include <stdlib.h>
+
+int stirling2(int n, int k){
+    if (k < 0 || k > n || n < 0){
+        return 0;
+    }
+    int arraySize = n + 1;
+    int *prev = (int*)calloc(arraySize, sizeof(int));
+    int *curr = (int*)calloc(arraySize, sizeof(int));
+    prev[0] = 1;
+    if(n > 0){
+        //outer for loop iterates over each row of triangle
+        for (int i=1; i<arraySize; i++){
+            // build up next row in curr
+            for (int j=1; j<arraySize; j++){
+                curr[j] = prev[j - 1];
+                curr[j] += prev[j] * j;
+            }
```

Comment:
```python

import numpy as np


def stirling2(n, k):
    if k < 0 or k > n or n < 0:
        return 0
    if k <= n - k + 1:
        prev = [1]*k
        current = [None]*k
        for i in range(1, n - k + 1):
            current[0] = prev[0]
            for j in range(1, k):
                current[j] = (j+1) * prev[j] + current[j-1]
            prev = current
    else:
        prev = [1]*(n - k + 1)
        current = [None]*(n - k + 1)
        for i in range(1, k):
            current[0] = prev[0]
            for j in range(1, n - k + 1):
                current[j] = (i+1) * current[j-1] + prev[j]
            prev = current
    return prev[-1]
```

@rlucas7 Here's a Python implementation that only computes the values in the triangle that are actually needed. It should be relatively straightforward to translate into C. Note that we will want `malloc` instead of `calloc` now because we no longer initialize with zero entries.

Reply:
I've just updated the implementation above to only use one array instead of two.

Reply:
> We're doing too much work here and allocating more than we need to. To compute {nk} we don't need the full triangle. ... If k≤n−k+1 we can allocate two arrays of size k and compute the values in the parallelogram left to right and then top to bottom. If k>n−k+1 we can allocate two arrays of size n−k+1 and compute the values top to bottom and then left to right.

I understand this as taking lhs or the top (angular) array of `1`s in the Stirling triangle and proceeding to do the calculations these ways, along angled movement of a row. 

Clever. 

> This also has the benefit of making it so that if an overflow occurs at any time during execution, then the final result will overflow.

I agree. 

> If we compute the full triangle this won't be the case. We need to be able to detect overflow, and this will simplify that.

I don't follow the reasoning of these last 2 sentences. It seems to me that either case it's relatively straightforward to check. Unless I'm mistaken the same `n,k` values overflow in either case (using array based dynamic programs). Did I miss something?

Reply:
> I don't follow the reasoning of these last 2 sentences. It seems to me that either case it's relatively straightforward to check. Unless I'm mistaken the same n,k values overflow in either case (using array based dynamic programs). Did I miss something?

I just mean, in the reduced schedule where we only compute stirling numbers over the parallelogram, we just need to check if overflow occurs since every value computed will take part in the final result. If we do the whole triangle up to ${n \brace k\}$, some of the entries of the array will overflow but these won't be used in the final result. It's not that difficult to check with the full triangle, it's just that we need to check for overflow and also that the value will be used in the final result. I like the simpler condition where we just have to check for overflow better.

Reply:
> ... it's just that we need to check for overflow and also that the value will be used in the final result. I like the simpler condition where we just have to check for overflow better.

Got it-Thanks.

Reply:
FWIW After switching to long outputs in C (on my mac) the overflow bumps up from first happening at 17-18 to 25-26,
```
stirling2(25,0) = 0 
stirling2(25,1) = 1 
stirling2(25,2) = 16777215 
stirling2(25,3) = 141197991025 
stirling2(25,4) = 46771289738810 
stirling2(25,5) = 2436684974110751 
stirling2(25,6) = 37026417000002430 
stirling2(25,7) = 227832482998716310 
stirling2(25,8) = 690223721118368580 
stirling2(25,9) = 1167921451092973005 
stirling2(25,10) = 1203163392175387500 
stirling2(25,11) = 802355904438462660 
stirling2(25,12) = 362262620784874680 
stirling2(25,13) = 114485073343744260 
stirling2(25,14) = 25958110360896000 
stirling2(25,15) = 4299394655347200 
stirling2(25,16) = 526655161695960 
stirling2(25,17) = 48063331393110 
stirling2(25,18) = 3275678594925 
stirling2(25,19) = 166218969675 
stirling2(25,20) = 6220194750 
stirling2(25,21) = 168519505 
stirling2(25,22) = 3200450 
stirling2(25,23) = 40250 
stirling2(25,24) = 300 
stirling2(25,25) = 1 
stirling2(25,26) = 0 

stirling2(26,0) = 0 
stirling2(26,1) = 1 
stirling2(26,2) = 33554431 
stirling2(26,3) = 423610750290 
stirling2(26,4) = 187226356946265 
stirling2(26,5) = 12230196160292565 
stirling2(26,6) = 224595186974125331 
stirling2(26,7) = 1631853797991016600 
stirling2(26,8) = 5749622251945664950 
stirling2(26,9) = -7245227292754425991 
stirling2(26,10) = -5247188700862703611 
stirling2(26,11) = -8417665732711074856 
stirling2(26,12) = 5149507353856958820 
stirling2(26,13) = 1850568574253550060 
stirling2(26,14) = 477898618396288260 
stirling2(26,15) = 90449030191104000 
stirling2(26,16) = 12725877242482560 
stirling2(26,17) = 1343731795378830 
stirling2(26,18) = 107025546101760 
stirling2(26,19) = 6433839018750 
stirling2(26,20) = 290622864675 
stirling2(26,21) = 9759104355 
stirling2(26,22) = 238929405 
stirling2(26,23) = 4126200 
stirling2(26,24) = 47450 
stirling2(26,25) = 325 
stirling2(26,26) = 1 
stirling2(26,27) = 0 
```